ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+#import "ADKeychainUtil.h"
+#import "ADLogger+Internal.h"
+#import "ADWorkPlaceJoinConstants.h"
+#import "ADRegistrationInformation.h"
+
+// Convenience macro for checking keychain status codes while looking up the WPJ information.
+#define CHECK_KEYCHAIN_STATUS(_operation) \
+{ \
+  if (status != noErr) \
+  { \
+    ADAuthenticationError* adError = [ADAuthenticationError keychainErrorFromOperation:_operation status:status correlationId:correlationId];\
+    if (error) { *error = adError; } \
+    goto _error; \
+  } \
+}
+
 
 @implementation ADWorkPlaceJoinUtil
 
-+ (ADRegistrationInformation*)getRegistrationInformation:(NSUUID *)correlationId
-                                                   error:(ADAuthenticationError * __autoreleasing *)error
++ (ADRegistrationInformation*) getRegistrationInformation:(NSUUID*)correlationId
+                                                    error:(ADAuthenticationError * __autoreleasing *)error
+{
+    ADRegistrationInformation* info = nil;
+    SecIdentityRef identity = nil;
+    SecCertificateRef certificate = nil;
+    SecKeyRef privateKey = nil;
+    NSString *certificateSubject = nil;
+    NSData *certificateData = nil;
+    NSString *userPrincipalName = nil;
+    NSString* certificateIssuer  = nil;
+    ADAuthenticationError* adError = nil;
+    
+    if (error)
+        *error = nil;
+    
+    AD_LOG_VERBOSE(@"Attempting to get WPJ registration information", correlationId, nil);
+    
+    [self copyCertificate:correlationId identity:&identity certificate:&certificate certificateIssuer:&certificateIssuer error:&adError];
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ certificate and identify - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    // If there's no certificate in the keychain, return nil. adError won't be set if the
+    // cert can't be found since this isn't considered an error condition.
+    if (!certificate)
+        return nil;
+    
+    certificateSubject = (__bridge_transfer NSString*)(SecCertificateCopySubjectSummary(certificate));
+    certificateData = (__bridge_transfer NSData*)(SecCertificateCopyData(certificate));
+    
+    
+    // Get the private key
+    AD_LOG_VERBOSE(@"Retrieving WPJ private key reference", correlationId, nil);
+    
+    privateKey = [self copyPrivateKeyRefForIdentifier:privateKeyIdentifier correlationId:correlationId error:&adError];
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ private key reference - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    
+    // Get user principal name
+    AD_LOG_VERBOSE(@"Retrieving WPJ user principal name", correlationId, nil);
+    
+    userPrincipalName = [self stringDataFromIdentifier:upnIdentifier correlationId:correlationId error:&adError];
```

Comment:
Yes, and this is a constant source of problems for the WPJ library and a limitation that will need to be addressed if we ever support multiple WPJs on a device. Considering we're starting anew on macOS with Conditional Access we should not repeat the same mistakes.